pkg/beacon/relay/gjkr/protocol.go

```diff
@@ -1643,20 +1657,69 @@ func pow(id group.MemberIndex, y int) *big.Int {
 //    public keys were reconstructed in Phase 11.
 //
 // See Phase 12 of the protocol specification.
-func (rm *CombiningMember) CombineGroupPublicKey() {
+func (cm *CombiningMember) CombineGroupPublicKey() {
 	// Current member's individual public key `A_i0`.
-	groupPublicKey := rm.individualPublicKey()
+	groupPublicKey := cm.individualPublicKey()
 
 	// Add received peer group members' individual public keys `A_j0`.
-	for _, peerPublicKey := range rm.receivedValidPeerIndividualPublicKeys() {
+	for _, peerPublicKey := range cm.receivedValidPeerIndividualPublicKeys() {
 		groupPublicKey = new(bn256.G2).Add(groupPublicKey, peerPublicKey)
 	}
 
 	// Add reconstructed misbehaved members' individual public keys `G * z_m`.
-	for _, peerPublicKey := range rm.reconstructedIndividualPublicKeys {
+	for _, peerPublicKey := range cm.reconstructedIndividualPublicKeys {
 		groupPublicKey = new(bn256.G2).Add(groupPublicKey, peerPublicKey)
-
 	}
 
-	rm.groupPublicKey = groupPublicKey
+	cm.groupPublicKey = groupPublicKey
+}
+
+// ComputeGroupPublicKeyShares computes group public key shares for each
+// individual member in the group. Those group public key shares are
+// needed to perform the verification of relay entry signature shares coming
+// from given group member.
+func (cm *CombiningMember) ComputeGroupPublicKeyShares() {
+	go func() {
+		groupPublicKeyShares := make(map[group.MemberIndex]*bn256.G2)
+
+		// Calculate group public key shares for all other operating members.
+		for _, operatingMemberID := range cm.group.OperatingMemberIDs() {
+			if operatingMemberID == cm.ID {
+				continue
+			}
+
+			// Calculate the first public key share for the given operating
+			// member based on the current member public key share points.
+			sum := cm.publicKeyShare(operatingMemberID, cm.publicKeySharePoints)
+
+			// Iterate through the `QUAL` set and calculate subsequent
+			// public key share for the given operating member based on...
+			for qualifiedMemberID := range cm.receivedQualifiedSharesS {
+				// ...received and valid member's public key share points...
+				if publicKeySharePoints, ok := cm.receivedValidPeerPublicKeySharePoints[qualifiedMemberID]; ok {
+					publicKeyShare := cm.publicKeyShare(
+						operatingMemberID,
+						publicKeySharePoints,
+					)
+					sum = new(bn256.G2).Add(sum, publicKeyShare)
+					// ...OR in case given sender didn't send their public key
+					// share points, take their reconstructed share and recover
+					// the public key share.
+				} else {
+					for _, shares := range cm.revealedMisbehavedMembersShares {
+						if shares.misbehavedMemberID == qualifiedMemberID {
+							publicKeyShare := new(bn256.G2).ScalarBaseMult(
+								shares.peerSharesS[operatingMemberID],
+							)
+							sum = new(bn256.G2).Add(sum, publicKeyShare)
+						}
+					}
+				}
+			}
+
+			groupPublicKeyShares[operatingMemberID] = sum
+		}
```

Comment:
I think it's worth wrapping the code in this goroutine with an info log with a classic `[member:x] ...` pattern informing when do we start computations and when they are completed.

Reply:
Addressed in #1547.